graphql/websocket.go

```diff
@@ -139,20 +138,9 @@ func (w *webSocketClient) forwardWebSocketData(message []byte) error {
 	if wsMsg.ID == "" { // e.g. keep-alive messages
 		return nil
 	}
-	w.subscriptions.Lock()
-	defer w.subscriptions.Unlock()
-	sub, success := w.subscriptions.map_[wsMsg.ID]
-	if !success {
-		return fmt.Errorf("received message for unknown subscription ID '%s'", wsMsg.ID)
-	}
-	if sub.hasBeenUnsubscribed {
-		return nil
-	}
-	if wsMsg.Type == webSocketTypeComplete {
-		sub.hasBeenUnsubscribed = true
-		w.subscriptions.map_[wsMsg.ID] = sub
-		reflect.ValueOf(sub.interfaceChan).Close()
-		return nil
+	sub, err := w.subscriptions.GetOrHandleCompletion(wsMsg.ID, wsMsg.Type)
```

Comment:
Move this into a function so that the lock is not kept when doing `sub.forwardDataFunc`. Hopefully this will prevent the deadlock.

Reply:
Scratch that, this made it worse!